src/Menu/Menu.js

```diff
@@ -0,0 +1,108 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* @flow */
+import React from 'react';
+import { createStyledComponent } from '../utils';
+import { MenuDivider, MenuGroup, MenuItem } from './index';
+
+type Props = {
+  /** [MenuDivider](./menu-divider), [MenuGroup](./menu-group), or [MenuItem](./menu-item) */
+  children?: MnrlReactNode,
+  /** Data used to contruct Menu. See [example](#data) */
+  data?: Array<{ items: Array<Item>, title?: MnrlReactNode }>,
+  /** @Private Function that returns props to be applied to each item */
+  getItemProps?: (props: Object, scope: Object) => Object
+};
+
+type Item = {
+  iconEnd?: React$Element<*>,
+  iconStart?: React$Element<*>,
+  disabled?: boolean,
+  divider?: boolean,
+  onClick?: (event: Object) => void,
+  render?: (item: Object, props: Object, theme: Object) => React$Element<*>,
+  secondaryText?: MnrlReactNode,
+  text?: MnrlReactNode,
+  variant?: 'regular' | 'danger' | 'success' | 'warning'
+};
+
+const Root = createStyledComponent(
+  'div',
+  {},
+  {
+    displayName: 'Menu',
+    includeStyleReset: true
+  }
+);
+
+/**
+ * Menus display a list of [MenuItems](./menu-item). The items can have actions or be used for navigation.
+ */
+export default function Menu({
+  children,
+  data,
+  getItemProps,
+  ...restProps
+}: Props) {
+  const rootProps = {
+    ...restProps
+  };
+
+  return (
+    <Root {...rootProps}>
+      {Array.isArray(data)
+        ? data.reduce(
+            (acc, group, groupIndex) => {
+              acc.groups.push(
+                <MenuGroup key={groupIndex} title={group.title}>
+                  {group.items.map((item, itemIndex) => {
+                    if (item.divider) {
+                      return <MenuDivider key={itemIndex} />;
+                    } else {
+                      const index = acc.itemIndex++;
+                      const { text, ...restItemProps } = item;
+                      const itemProps = getItemProps
+                        ? getItemProps(
+                            {
+                              ...restItemProps,
+                              item
+                            },
+                            { index, item }
+                          )
+                        : {
+                            ...restItemProps,
+                            item
+                          };
+
+                      return (
+                        <MenuItem key={itemIndex} {...itemProps}>
+                          {text}
+                        </MenuItem>
+                      );
+                    }
+                  })}
+                </MenuGroup>
+              );
+
+              return acc;
+            },
+            { groups: [], itemIndex: 0 }
```

Comment:
oof. would all this fit into a named function(s)? Just thinking about our (hopeful) future contributors.

Reply:
Updated to hopefully improve readability.